src/Lean/Compiler/IR/EmitLLVM.lean

```diff
@@ -1544,7 +1564,19 @@ def emitLLVM (env : Environment) (modName : Name) (filepath : String) (tripleStr
   | .ok _ => do
          let membuf ← LLVM.createMemoryBufferWithContentsOfFile (← getLeanHBcPath).toString
          let modruntime ← LLVM.parseBitcode llvmctx membuf
+         let runtime_global_names ← getModuleGlobals modruntime
+         let runtime_function_names ← getModuleFunctions modruntime
          LLVM.linkModules (dest := emitLLVMCtx.llvmmodule) (src := modruntime)
+         -- Mark every global and function as having internal linkage.
+         for name in runtime_global_names do
+           let some global ← LLVM.getNamedGlobal emitLLVMCtx.llvmmodule name
+              | throw <| IO.Error.userError s!"ERROR: linked module must have global from runtime module: '{name}'"
+           LLVM.setLinkage global LLVM.Linkage.internal
+         for name in runtime_function_names do
+           let some fn ← LLVM.getNamedFunction emitLLVMCtx.llvmmodule name
+              | throw <| IO.Error.userError s!"ERROR: linked module must have function from runtime module: '{name}'"
+           LLVM.setLinkage fn LLVM.Linkage.internal
+
          optimizeLLVMModule emitLLVMCtx.llvmmodule
          LLVM.writeBitcodeToFile emitLLVMCtx.llvmmodule filepath
          let tripleStr := tripleStr?.getD (← LLVM.getDefaultTargetTriple)
```

Comment:
I am curious, but was wondering if these changes + the stage0 update could be factored out into a separate PR?

Reply:
Yes, these probably should have been part of #2410

Reply:
Before we pull this out I would be curious to understand as to why this PR is failing know unlike #2340 :(

Reply:
From diffing this against the newly updated #2340 it seems to me as if there is only diff in the update stage0 due to the different times that these were done. I don't understand how this ends us up with a failing linker in the LLVM CI though.

Reply:
I'm not sure what you mwean by "different times", this branch does not have an update-stage0 at all. And it definitely needs one exactly for the linker fix it contains and itself needs.

Reply:
Right. I chatted about this with @bollu, I feel trying to debug these differences is a time-sink that we can avoid by pulling stuff out of #2340 and keeping it updated. I am trying to do this at the moment.